### ISSUE TITLE
Support for COS - ActivityTracker integration and IP filtering configuration

### ISSUE THREAD

**Issue by ifs-pauljegouic** [IBM-Cloud/terraform-provider-ibm#1487]
Hi there,

We would like to be able to for each bucket to:

- Configure the activity_tracker_crn to integrate with,
- Configure the IP filtering.

though the Terraform resource.

For now, we are using a null_resource that uses COS SDK.

**Comment by ifs-pauljegouic** [comment]
Hello there, there is an additional need regarding this. Let me know if you want me to create another ticket.

- From the console I am still able to configure the Bucket AccessPolicies (that includes the IP Firewall)
- With Terraform, on the refresh phase, I got a 403 Forbidden error.

It should not be.

@hkantare

**Comment by hkantare** [comment]
```
    Configure the activity_tracker_crn to integrate with,
    Configure the IP filtering.
```

We are working on this feature to attach firewall ip's activity tracker, monitoring crn...It will be part of the next release

**Comment by ifs-pauljegouic** [comment]
Great !

**Comment by hkantare** [comment]
@ifs-pauljegouic Regarding this `With Terraform, on the refresh phase, I got a 403 Forbidden error.`

Since you added some Bucket AccessPolicies (that includes the IP Firewall) did you added your systemIP to the list of allowed IP's  or your system IP fall in the range of IP address...Until you add it you can't access the buckets

<img width="1680" alt="Screenshot 2020-06-05 at 3 25 23 PM" src="https://user-images.githubusercontent.com/24930306/83866523-8d188d80-a745-11ea-978a-a27795235075.png">

**Comment by hkantare** [comment]
The above comments are based on our investigation we found it

**Comment by hkantare** [comment]
<img width="857" alt="Screen Shot 2020-06-05 at 4 01 03 PM" src="https://user-images.githubusercontent.com/24930306/83866706-d23cbf80-a745-11ea-8f69-95011d04fdcc.png">

**Comment by ifs-pauljegouic** [comment]
ARGH so even the properties (including IDs ?) are handled on the data plane of COS ...

So we don't have any solution on this use case ?

- Provision a Bucket with AccessPolicies
- Manage those AccessPolicies from Terraform (to keep the track of configurations) from another location than the ones informed in AccessPolicies (I am performing my Terraform execution from a CICD with dynamic IPs ... :'()

I mean no easy solution, it implies SSH tunnel or VPN.

I asked the COS team to distinguish more the "administration of the service" and the "usage of the service".

**Comment by ifs-pauljegouic** [comment]
Mmmmm @hkantare I need to make more tests. 

Yesterday I SWEAR, I was able to edit the access policies from the console even if I was doing this from another IP than the ones informed in "Authorized IPs" 

And today, I am not able.

**Comment by ifs-pauljegouic** [comment]
Oh wow. After some investigations, I think I will have to make a ticket support ...

Here is my usecase:

1/ I provision a COS with a Bucket
2/ I inform an AccessPolicy with an IP 192.168.0.0/16 (which is not my SystemIP)
3/ I try to navigate under Configuration Bucket --> Denied OK
4/ I try to navigate under Object --> Denied OK
5/ I go back to the AccessPolicies menu, I can see the IPs allowed, and still edit them.
6/ I add one.
7/ I leave the Bucket Menu.
8/ --> Bucket is totally denied.

I am going to check if the second IP I have informed in step 6 has been taken into account. That would result in a security breach.

![usecase](https://user-images.githubusercontent.com/50360031/83868556-62214000-a72b-11ea-94af-16002906ac09.gif)

**Comment by ifs-pauljegouic** [comment]
After some investigations thanks to @kevinisninja from GCAT, we have those conclusions:

- Access Policies are still editable after a IP firewall apply. But there's a trick. (see the GIF)

--> This leads me to the initial situation where COS has 2 endpoints, and It would be great that even if the Bucket content should not be access by an IP, it can still be managed through Terraform without getting a 403 error :) 

If the console can do it, I bet Terraform teams can do it better than the GUI Team ;)

For now:
- AccessPolicies seems to be possible
- Configuration is not.

![usecase2](https://user-images.githubusercontent.com/50360031/83899771-47fe5680-a759-11ea-9d5f-48032f5e160d.gif)

**Comment by kevinisninja** [comment]
Hey @hkantare, I think this might need to be pushed internally, and might be outside the scope of the provider.

The customer wants terraform to be able to refresh the state of the bucket properly, without whitelisting management infrastructure location IP

AFAIK This separation of control plane and data plane is possible today in ICD databases, where adding an IP whitelist will prevent access to the databases in the instance, but terraform is still able to configure/refresh the state of the instance

**Comment by kevinisninja** [comment]
@hkantare Do you know what exactly what attributes is the terraform provider trying to get that obtains a 403 failure?

Because when the bucket has an IP whitelist, when you go view it via the UI, you can still see the COS instance, as well as all the buckets that are in there. What information is the provider trying to obtain that causes it to fail, since it seems like you have all the information necessary in the UI. 

Even with the IP firewall applied, you can still see the buckets that exist, so it doesn't seem like the provider should be failing, even if the system IP is not IP whitelisted.

**Comment by hkantare** [comment]
@kevinisninja terraform provider will try to set some of the bucket information like location, crn,..so on
https://github.com/IBM-Cloud/terraform-provider-ibm/blob/0657d8cd97173dd8c63278d0c90c86de1224e864/ibm/resource_ibm_cos_bucket.go#L121

Here you can find the set of code which we call to get the bucket information

**Comment by kevinisninja** [comment]
@hkantare per the information @ifs-pauljegouic provided, 

the GUI is still able to get information regarding the buckets, even though he cannot access one of the buckets because his system ip is not whitelisted in the ip firewall. 

so is there some other API with the COS team that they are using in the GUI that allows them to get that information, that maybe the provider should be using instead?

**Comment by hkantare** [comment]
@kevinisninja The provider is using the https://github.com/IBM/ibm-cos-sdk-go/ below SDK to make calls to the API

**Comment by hkantare** [comment]
@kevinisninja This is the call failing when we add a firewall IP with 403 error code https://github.com/IBM-Cloud/terraform-provider-ibm/blob/0657d8cd97173dd8c63278d0c90c86de1224e864/ibm/resource_ibm_cos_bucket.go#L166

**Comment by kevinisninja** [comment]
@hkantare I talked with the COS team. They said that instead of using GetBucketLocation, you should be able to use ListBucketsExtendedOutput instead, and it will be able to provide you that information despite the IP firewall. 
https://ibm.github.io/ibm-cos-sdk-go/service/s3/ListBucketsExtendedOutput.html

From their perspective, the bucket list is supposed to be visible, only Object and Bucket operations should be blocked.

Is it possible to refactor the provider code so that there will no longer be a 403 error?

**Comment by hkantare** [comment]
We changed to use ListBucketsExtendedOutput as per above comments.We are no longer getting 403 error now

The fix is available in master will be part of next release

**Comment by ifs-pauljegouic** [comment]
Great !

**Comment by hkantare** [comment]
Fixed in latest release https://github.com/IBM-Cloud/terraform-provider-ibm/releases/tag/v1.8.0
https://github.com/IBM-Cloud/terraform-provider-ibm/blob/master/website/docs/r/cos_bucket.html.markdown

Cloud docs Docs will be updated shortly

**Comment by hkantare** [comment]
Available in latest releases

**Comment by ifs-pauljegouic** [comment]
Does not work properly

**Comment by hkantare** [comment]
Can you explain the scenario or the steps in which its not working properly

**Comment by ifs-pauljegouic** [comment]
4sure. I am writing to Support because stuff have evolved in opposite direction between you and COS team.

Loosing patience with IBM coordination

**Comment by umarali-nagoor** [comment]
Hi Pauljegouic,

Executed following scenario to test the “COS bucket activity tracker and ip filtering configuration” CRUD functionalities and its observed that these operations are working as expected. 

Here are the steps and configurations that we have executed

Steps Followed: 

1.  Created VM Instance (say virtual machine with public ip 169.60.175.87 )

2.  In local (mac) machine,  include public ip of VM instance ( created in step-1 ) as part of allowed_ip list and create a COS bucket instance along with activity tracker 
		
		allowed_ip =  [“169.60.175.87”]

3.  Since VM instance public ip is part of white-list ip list, we should be able to perform update or delete operations from ip 
     169.60.175.87

4.  To do so, copied the terraform tf file and state file  ( which was used to create the COS bucket  in local machine ) from my local machine to VM instance

5.   In Virtual machine, update activity tracker and allowed_ip configurations and run “terraform apply”

6.  Able to update the same and verified from GUI side as well

7.   Now try to destroy the COS bucket in VM instance by running “terraform destroy”, even the step is going through good        

Configuration tf file

resource "ibm_resource_instance" "cos_instance" {
  name              = "cos-instance"
  resource_group_id = data.ibm_resource_group.cos_group.id
  service           = "cloud-object-storage"
  plan              = "standard"
  location          = "global"
}

resource "ibm_resource_instance" "activity_tracker" {
  name              = "activity_tracker"
  resource_group_id = data.ibm_resource_group.cos_group.id
  service           = "logdnaat"
  plan              = "lite"
  location          = "us-south"
}

resource "ibm_resource_instance" "metrics_monitor" {
  name              = "metrics_monitor"
  resource_group_id = data.ibm_resource_group.cos_group.id
  service           = "sysdig-monitor"
  plan              = "lite"
  location          = "us-south"
}

resource "ibm_cos_bucket" "standard-ams03" {
  bucket_name          = "my1bucket-mine012090"
  resource_instance_id = ibm_resource_instance.cos_instance.id
  cross_region_location      = "us"
  storage_class        = "standard"

  activity_tracking {
    read_data_events     = true
    write_data_events    = true
    activity_tracker_crn = ibm_resource_instance.activity_tracker.id
  }

  metrics_monitoring {
    usage_metrics_enabled  = true
    metrics_monitoring_crn = ibm_resource_instance.metrics_monitor.id
  }

  allowed_ip =  [“169.60.175.87”]
 
}

Can you please share the steps that you executed. Thanks in advance.

**Comment by ifs-pauljegouic** [comment]
In fact, here is the need:

- We are managing, administrating COS resource from our CICD pipeline,
- We are accessing the content of the bucket through the VPC (and we want to filter inbound with CSE VPC IPs)

Right now, I understand, that it is not possible. I should have had a meeting this afternoon with COS Team, but it was cancelled. I agree that the asked capability cannot be implemented by the TF Team only.

+ followings.

**Try to import an existing bucket**

Try to redo the same process:

```hcl
terraform init
terraform plan 
terraform apply
```

You will noticed that the plan does not see any changes.

As my team is used to Terraform, this mean, that the configuration retrieved by TF Provider, is the same as the one informed in the input. So we thought that the behaviour expected was implemented.

But in fact, it seems that the initial value is stored in the state, but is not refreshed with the actual configuration.

**Try to import an existing bucket**

Now perform `terraform state rm resource.ibm_cos_bucket.standard-ams03` and `terraform import  resource.ibm_cos_bucket.standard-ams03 EXISTING_CRN`

From your local machine (which acts with configuration as our pipeline - not whitelisted), and it will fail. This prove that the `sysdig`and `activity_tracker` are not refreshed.

In conclusion, this issue has been partially solved. (I think that all the misunderstood comes from the fact that this issue thread actually contains 2 issues ...)
We will discuss with the COS team, so that they can provide a solution on how to differentiate administration endpoints and usage endpoints (We would like the same behaviour regarding data & control as for ICD in fact).

**Comment by hkantare** [comment]
We are now closing the issue for now may be once we have updates from COS team you can open a new issue if it requires any changes from Terraform end..
